[COR-17] Build Golang SDK

rukolahasser opened the pull request:
Goal: 
1. Provide more flexibility to devs when building connectors.
2. Devs make a minimal effort to build connectors.

Changes:
1. Refactor `connector`, move chain-specific clients to separate modules.
3. Create chain-specific base connectors.
4. Create a working example of connectors that use the latest golang SDK

azhang commented on examples/README.md:
```diff
@@ -0,0 +1,7 @@
+# Example Connectors
```
# Example Connectors

The examples in this directory show various ways to get started with building Nakji connectors.

- The Ethereum example uses the base Nakji connector library to send blocks and transactions to Nakji.
- The Compound example uses the [ethereum chain-specific connector library](../chain) to send ethereum events to Nakji.

These connector examples in this directory are simplified and commented, and may not handle everything the production connector does. The goal is just to demonstrate the concept of a connector for those who like to read code over reading documentation.

For those who want to see full connectors that are actually running in production as additional examples, please refer to github.com/nakji-network/connectors.

azhang commented on examples/ethereum/cmd/ethereum/main.go:
```diff
@@ -2,31 +2,10 @@
-	}
-
-	ethConnector.Start()
+	ec := ethereum.NewConnector()
```
Can we keep

```
	// Any additional command line arguments, such as chain selection override. Set up via https://pkg.go.dev/github.com/spf13/viper#readme-working-with-flags
		// Chain: "bsc",

		// Any additional config vars from the config yaml, using functions from Viper (https://pkg.go.dev/github.com/spf13/viper#readme-getting-values-from-viper)
		// This is namespaced via connector id (author-name-version)
		// CustomOption: c.Config.GetString("custom_option"),
		```

azhang commented on examples/ethereum/ethereum.go:
```diff
@@ -42,31 +81,26 @@ func (c *EthereumConnector) Start() {
-					log.Error().Err(err).Msg("Kafka write proto")
-				}
+
+				c.EventSink <- buildKafkaMsg(kafkautils.MsgTypeFct, &blockData)
```
Let's use transactions and function calls to send kafka messages instead of chan

rukolahasser replied:
I think it'd be better to let `connector` lib handle the kafka interaction, so they wouldn't use it in the wrong way.

azhang reviewed (changes requested):
Added some comments.

azhang commented:
Regarding connector.go, do we want to use this PR to add a "ProduceMessages()" function that puts produces many messages in a single transaction?

cnkarz reviewed (commented):
- Looks good in general. Just wondering you removed some error handling.
- Maybe you can also mention that `ethereum` can also be used for evm chains like BSC.
- Also, it would be nice to have `WS()` and `HTTP()` getters from chain clients, so that we can query for historical data thru REST. But this can be part of another ticket.

rukolahasser commented on chain/ethereum/ethereum.go:
```diff
@@ -98,40 +109,25 @@ func ChunkedSubscribeFilterLogs(
 	}
 
 	if len(addresses) > addressChunkSize {
-		failedQueries, err = ChunkedFilterLogs(ctx, client, addresses[:addressChunkSize], fromBlock, toBlock, logChan, failedQueries)
```
I'll change it back

rukolahasser commented:
> Regarding connector.go, do we want to use this PR to add a "ProduceMessages()" function that puts produces many messages in a single transaction?

No. It will be in a separate task.

rukolahasser commented on chain/ethereum/ethereum.go:
```diff
@@ -145,36 +141,22 @@ func ChunkedFilterLogs(
 
 	logs, err := client.FilterLogs(ctx, query)
 	if err != nil {
-		if strings.Contains(err.Error(), "read limit") {
```
Added it back

rukolahasser commented on chain/ethereum/ethereum.go:
```diff
@@ -145,36 +141,22 @@ func ChunkedFilterLogs(
 
 	if fromBlock == toBlock {
-		return nil
+		fmt.Println("returning backifll ")
```
Added it back

rukolahasser commented:
> * Looks good in general. Just wondering you removed some error handling.
> * Maybe you can also mention that `ethereum` can also be used for evm chains like BSC.
> * Also, it would be nice to have `WS()` and `HTTP()` getters from chain clients, so that we can query for historical data thru REST. But this can be part of another ticket.

Added those error handling back. Also added some comments to tell users that it works for evm chains.

azhang commented on examples/ethereum/cmd/ethereum/main.go:
```diff
@@ -2,31 +2,10 @@
-
-		// Any additional config vars from the config yaml, using functions from Viper (https://pkg.go.dev/github.com/spf13/viper#readme-getting-values-from-viper)
-		// This is namespaced via connector id (author-name-version)
-		// CustomOption: c.Config.GetString("custom_option"),
```
at least the `chain` option should be defined here, because if it's defined in ethereum/ethereum.go then it's hardcoded in the library and then the library can't be reused for other chains

azhang commented:
> Also, it would be nice to have WS() and HTTP() getters from chain clients, so that we can query for historical data thru REST. But this can be part of another ticket.

Not sure if this is needed. users already have the option to call functions from the chain clients directly; I can't see how adding another layer of abstraction will simplify things and only increase the burden of our documentation / maintenance when chain clients update to incompatible versions @cnkarz

cnkarz commented:
> > Also, it would be nice to have WS() and HTTP() getters from chain clients, so that we can query for historical data thru REST. But this can be part of another ticket.
> 
> Not sure if this is needed. users already have the option to call functions from the chain clients directly; I can't see how adding another layer of abstraction will simplify things and only increase the burden of our documentation / maintenance when chain clients update to incompatible versions @cnkarz

During backfill, we query the ws connection too much that they break the connection. Then the connector crashes and restarts. Having a HTTP handle would allow us to use REST and not disturb the ws connection.

rukolahasser commented:
> > > Also, it would be nice to have WS() and HTTP() getters from chain clients, so that we can query for historical data thru REST. But this can be part of another ticket.
> > 
> > 
> > Not sure if this is needed. users already have the option to call functions from the chain clients directly; I can't see how adding another layer of abstraction will simplify things and only increase the burden of our documentation / maintenance when chain clients update to incompatible versions @cnkarz
> 
> During backfill, we query the ws connection too much that they break the connection. Then the connector crashes and restarts. Having a HTTP handle would allow us to use REST and not disturb the ws connection.
